### PR TITLE
Copy command

### PR DESCRIPTION
**Pull request by b0r6**
`./openvdc copy [filepath] [instanceID]:[path]`

As of now this can only copy a local file and transfer it to the instance machine, not the other way around. Ideally one should also be able to copy files between instances as well, but I'll implement that at some later point.

### CODE REVIEW

**Review by unakatsuo — CHANGES_REQUESTED** [review]
missing ``go fmt``.

File: cmd/openvdc/cmd/copy.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
+package cmd
+
+import (
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/axsh/openvdc/api"
+	filecopy "github.com/axsh/openvdc/cmd/openvdc/cmd/copy"
```
Comment (unakatsuo) [inline cmd/openvdc/cmd/copy.go:8]: weird namespace.

File: cmd/openvdc/cmd/copy/copy.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		return true
+	}
+}
+
+func (c *Client) CopyFile(filePath string, instanceDir string) error {
+	if !fileExists(filePath) {
+		return errors.New("Source file not found")
+	}
+
+	file, _ := os.Open(filePath)
```
Comment (unakatsuo) [inline cmd/openvdc/cmd/copy/copy.go]: Please do not squash the error from IO operation.

File: cmd/openvdc/cmd/copy/copy.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	session, err := cl.NewSession()
+	if err != nil {
+		return errors.Wrap(err, "Failed to create session")
+	}
+	defer session.Close()
+
+	go func() {
+		w, _ := session.StdinPipe()
+		defer w.Close()
+		fmt.Fprintln(w, "C0655", int64(len(b)), srcFilename)
```
Comment (unakatsuo) [inline cmd/openvdc/cmd/copy/copy.go]: Here can rewrite with https://golang.org/pkg/os/#File.Stat. No need the buffering the content at L57-58. It is scary for the large file.

File: cmd/openvdc/cmd/copy/copy.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+		io.Copy(w, br)
```
Comment (unakatsuo) [inline cmd/openvdc/cmd/copy/copy.go]: Just pass ``file`` at L54. And please do not forget to handle the error from the function.

```
io.Copy(w, file)
```

File: cmd/openvdc/cmd/copy/copy.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	if !fileExists(filePath) {
+		return errors.New("Source file not found")
+	}
+
+	file, _ := os.Open(filePath)
+	defer file.Close()
+
+	b, _ := ioutil.ReadAll(file)
+	br := bytes.NewReader(b)
+	srcFilename := path.Base(filePath)
```
Comment (unakatsuo) [inline cmd/openvdc/cmd/copy/copy.go]: ``path`` -> ``filepath``

File: cmd/openvdc/cmd/copy/copy.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		return false
+	} else {
+		return true
+	}
+}
+
+func (c *Client) CopyFile(filePath string, instanceDir string) error {
+	if !fileExists(filePath) {
+		return errors.New("Source file not found")
+	}
```
Comment (unakatsuo) [inline cmd/openvdc/cmd/copy/copy.go]: ``filesExists()`` is redundant.

```
file, err := os.Open(filePath)
if err != nil {
  if os.IsNotExist(err) {
    return errors.New("Source file not found")
  }
  return errors.Wrap(err, "Open failed")
}
defer file.Close()
```

**Review by unakatsuo — CHANGES_REQUESTED** [review]

**Review by unakatsuo — CHANGES_REQUESTED** [review]
I guess this change only works for LXC hypervisor. It should return error if the destination is the unsupported hypervisor type.

File: cmd/openvdc/cmd/copy.go (hunk 1 above)
Comment (b0r6) [reply cmd/openvdc/cmd/copy.go:8]: Ok. Any suggestion?

Comment (unakatsuo) [reply cmd/openvdc/cmd/copy.go:8]: I did same hack for console. please ignore.

File: cmd/openvdc/cmd/copy/copy.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+	session, err := cl.NewSession()
+	if err != nil {
+		return errors.Wrap(err, "Failed to create session")
+	}
+	defer session.Close()
+
+	go func() {
+		w, _ := session.StdinPipe()
```
Comment (unakatsuo) [inline cmd/openvdc/cmd/copy/copy.go]: Squashing error.

File: cmd/openvdc/cmd/copy/copy.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+		defer w.Close()
+		fmt.Fprintln(w, "C0655", int64(fileinfo.Size()), srcFilename)
+		io.Copy(w, file) //Todo: handle error
```
Comment (unakatsuo) [inline cmd/openvdc/cmd/copy/copy.go]: Please handle the error.

**Review by unakatsuo — CHANGES_REQUESTED** [review]

File: cmd/openvdc/cmd/copy/copy.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+		fmt.Fprintln(w, "\x00")
+	}()
+
+	session.Run("/usr/bin/scp -t " + instanceDir)
```
Comment (unakatsuo) [inline cmd/openvdc/cmd/copy/copy.go]: Squashing error.

**Review by unakatsuo — CHANGES_REQUESTED** [review]

**Comment by Metallion** [comment]
Looks good to me. If @unakatsuo  is happy with the changes, then I think this is good to go. :+1:

**Review by Metallion — APPROVED** [review]

File: cmd/openvdc/cmd/copy.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+		if err != nil {
+			log.WithError(err).Fatal("Disconnected abnormally")
+		}
+
+		client, err := filecopy.NewClient(res)
+		if err != nil {
+			log.WithError(err).Fatal("Failed to create client")
+		}
+
+		err = client.CopyFile(src, instanceDir)
```
Comment (unakatsuo) [inline cmd/openvdc/cmd/copy.go:68]: Does ``instanceDir`` point to the path on the host filesystem? It does not make sense. This feature is to copy the file into the instance filesystem not to the host filesystem.

**Review by unakatsuo — CHANGES_REQUESTED** [review]

File: cmd/openvdc/cmd/copy.go (hunk 10 above)
Comment (b0r6) [reply cmd/openvdc/cmd/copy.go:68]: Example:
```./openvdc copy testfile i-0000000001:/tmp/```

instanceDir refers to whichever folder you want to put the file in on the instance filesystem, in this case "/tmp/".

The file ends up in ```"/var/lib/lxc/i-0000000001/rootfs/tmp/"``` not ```"/tmp/"```

File: cmd/openvdc/cmd/copy/copy.go [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+	}
+
+	return &Client{
+		ClientConfig: config,
+		Host:         host,
+		Port:         port,
+	}, nil
+}
+
+func fileExists(path string) bool {
```
Comment (unakatsuo) [inline cmd/openvdc/cmd/copy/copy.go]: You can remove this unused function.

**Review by unakatsuo — CHANGES_REQUESTED** [review]

File: cmd/openvdc/cmd/copy.go (hunk 10 above)
Comment (unakatsuo) [reply cmd/openvdc/cmd/copy.go:68]: Miss read the code. ``/usr/bin/scp`` is run by executor's sshd so that data stream is sent to the container.

**Review by unakatsuo — APPROVED** [review]

**Pull request merged**